aten/src/ATen/FunctionalTensorImplBase.cpp

```diff
@@ -0,0 +1,184 @@
+    switch (view_meta.view_type) {
+      case ViewMeta::Type::view:
+          t = t.view(view_meta.size);
+          break;
+      case ViewMeta::Type::noOp:
+          break;
+      default:
+          TORCH_CHECK(false, "Tried to run the functionalization pass on an unsupported view: ", view_meta.view_type);
+    }
+  }
+  // We want the new tensor to have separate memory from the alias.
+  // Note that the clone required for functorch, but probably unnecessary for backends
+  // like vulkan and xla, because their view operators already create fresh tensors.
+  t = t.clone();
+  // This call to replace_() doesn't go through the dispatcher.
+  // It's a virtual method implemented directly on TensorImpl subclasses.
+  replace_(t.unsafeGetTensorImpl());
+  generation_ = alias_->generation();
+}
+
+bool FunctionalTensorImplBase::is_up_to_date() const {
+  if (alias_) {
+    return generation_ == alias_->generation();
+  }
+  return true;
+}
+
+void FunctionalTensorImplBase::maybe_add_update(const Tensor& updated_val) {
+  // If the mutated tensor wasn't a view, we don't need to do anything
+  if (is_view()) {
+    alias_->add_update(updated_val.clone(), view_metas_);
+    // TODO: add this back see if anything breaks?
+    //generation_ = alias_->generation(); // self is fully up to date at this point
+  }
+}
+
+void FunctionalTensorImplBase::set_view_meta(const Tensor& other, at::ViewMeta meta) {
+    auto other_impl = dynamic_cast<at::FunctionalTensorImplBase*>(other.unsafeGetTensorImpl());
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(other_impl != nullptr);
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(alias_ == nullptr);
+    if (other_impl->alias_ != nullptr) {
+        // if the other tensor is already a view, copy its ViewMeta vector and push the current one.
+        std::vector<at::ViewMeta> metas = other_impl->view_metas_;  // copy
+        metas.push_back(meta);
+        //alias_ = other_impl->alias_; // refcount bump
+        alias_ = other_impl->alias_; // refcount bump
+        view_metas_ = metas;
+    } else {
+        std::shared_ptr<Alias> alias = std::make_shared<Alias>(const_cast<Tensor&>(other));
+        at::ViewMeta noop_view_info(at::ViewMeta::Type::noOp, other.sizes().vec(), other.sizes().vec());
+        // The original tensor wasn't a view - turn it into a (no-op) view.
+        other_impl->alias_ = alias;
+        other_impl->view_metas_ = std::vector<at::ViewMeta>{noop_view_info};
+        // Turn the new tensor into a view too
+        alias_ = alias;
+        view_metas_ = std::vector<at::ViewMeta>{meta};
+    }
+}
+
+bool FunctionalTensorImplBase::is_view() const {
+    return view_metas_.size() != 0;
+}
+
+const char* FunctionalTensorImplBase::tensorimpl_type_name() const {
+    return "FunctionalTensorImplBase";
+}
+
+namespace functionalization {
+namespace impl {
+
+void maybe_add_update(Tensor& self) {
+  auto functional_base_impl = dynamic_cast<at::FunctionalTensorImplBase*>(self.unsafeGetTensorImpl());
```

Comment:
Shouldn't use dynamic cast, this is not friendly for platforms that don't have RTTI. Instead, test if the tensor has the Functionalize key, and if it does you can unsafe cast

Reply:
(invariant is, tensor has functionalize key if and only if it is FunctionalTensorImplBase)

Reply:
factor this into a helper function and use it all below

Reply:
I had no idea this was platform dependent, cool. Does that mean that we have no way to debug_assert invariants like that, aside from testing locally? (That `t->key_set().has(c10::DispatchKey::Functionalization)` <==> `isinstance(t, FunctionalTensorImplBase)*`)